README.txt

TYPO3 Flash MP3 Player using the free Wordpress Flash player from 1 Pixel Out.
http://www.1pixelout.net/code/audio-player-wordpress-plugin





ChangeLog
(add new changes on top of this file)

09-10-27 Waldo Sparding <waldo@familie-sparding.de>
	* New feature: width and height can now be set via constant editor
	* New feature: a list of mp3 files can be set in the plugins flexform

07-11-15 Peter Klein  <pmk@io.dk>

	* Fixed problem with javascript not working correctly in Safari.

07-10-03 Peter Klein  <pmk@io.dk>

	* Fixed problem where upload folder wasn't created during install.
	* Fixed problem with flexform value "minitems" was set to 1, preventing the user from deleting a MP3 once it was selected.

07-01-24 Peter Klein  <pmk@io.dk>

	* Initial release


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext: "pmkmp3player"
#
# Auto generated 17-11-2009 18:37
#
# Manual updates:
# Only the data in the array - anything else is removed by next write.
# "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'PMK MP3 Player',
	'description' => 'TYPO3 Flash MP3 Player using the free Wordpress Flash player from 1 Pixel Out.',
	'category' => 'plugin',
	'shy' => 0,
	'version' => '1.2.2',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'uploadfolder' => 1,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Peter Klein, Waldo Sparding',
	'author_email' => 'pmk@io.dk, waldo@familie-sparding.de',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:26:{s:9:"ChangeLog";s:4:"76f8";s:10:"README.txt";s:4:"b245";s:12:"ext_icon.gif";s:4:"63f3";s:17:"ext_localconf.php";s:4:"3e30";s:14:"ext_tables.php";s:4:"b135";s:19:"flexform_ds_pi1.xml";s:4:"7eac";s:13:"locallang.xml";s:4:"c1d7";s:16:"locallang_db.xml";s:4:"6504";s:14:"doc/manual.sxw";s:4:"5ecf";s:14:"pi1/ce_wiz.gif";s:4:"fb72";s:33:"pi1/class.tx_pmkmp3player_pi1.php";s:4:"adf3";s:41:"pi1/class.tx_pmkmp3player_pi1_wizicon.php";s:4:"6d99";s:13:"pi1/clear.gif";s:4:"cc11";s:17:"pi1/locallang.xml";s:4:"ce81";s:19:"res/buttonmask1.gif";s:4:"520b";s:19:"res/buttonmask2.gif";s:4:"938f";s:22:"res/expressInstall.swf";s:4:"204f";s:12:"res/logo.gif";s:4:"2f8a";s:11:"res/map.gif";s:4:"ff46";s:16:"res/mp3player.js";s:4:"0bd9";s:14:"res/player.swf";s:4:"7b87";s:16:"res/swfobject.js";s:4:"16e2";s:29:"static/advanced/constants.txt";s:4:"0010";s:25:"static/advanced/setup.txt";s:4:"8955";s:27:"static/simple/constants.txt";s:4:"3bf3";s:23:"static/simple/setup.txt";s:4:"e832";}',
	'suggests' => array(
	),
);

?>

ext_tables.php
<?php
if (!defined ('TYPO3_MODE')) 	die ('Access denied.');

t3lib_div::loadTCA('tt_content');
$TCA['tt_content']['types']['list']['subtypes_excludelist'][$_EXTKEY.'_pi1']='layout,pages,select_key';
$TCA['tt_content']['types']['list']['subtypes_addlist'][$_EXTKEY.'_pi1']='pi_flexform';

t3lib_extMgm::addPlugin(array('LLL:EXT:pmkmp3player/locallang_db.xml:tt_content.list_type_pi1', $_EXTKEY.'_pi1'),'list_type');

// add flexform definition
t3lib_extMgm::addPiFlexFormValue($_EXTKEY.'_pi1', 'FILE:EXT:'.$_EXTKEY.'/flexform_ds_pi1.xml');

t3lib_extMgm::addStaticFile($_EXTKEY,"static/simple/","Flash MP3 Player (Simple)");
t3lib_extMgm::addStaticFile($_EXTKEY,"static/advanced/","Flash MP3 Player (Advanced)");

if (TYPO3_MODE=="BE")	$TBE_MODULES_EXT["xMOD_db_new_content_el"]["addElClasses"]["tx_pmkmp3player_pi1_wizicon"] = t3lib_extMgm::extPath($_EXTKEY).'pi1/class.tx_pmkmp3player_pi1_wizicon.php';
?>

pi1/class.tx_pmkmp3player_pi1_wizicon.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2007-2009 Peter Klein <pmk@io.dk>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/




/**
 * Class that adds the wizard icon.
 *
 * @author	Peter Klein <pmk@io.dk>
 * @package	TYPO3
 * @subpackage	tx_pmkmp3player
 */
class tx_pmkmp3player_pi1_wizicon {

					/**
					 * Processing the wizard items array
					 *
					 * @param	array		$wizardItems: The wizard items
					 * @return	Modified array with wizard items
					 */
					function proc($wizardItems)	{
						global $LANG;

						$LL = $this->includeLocalLang();

						$wizardItems['plugins_tx_pmkmp3player_pi1'] = array(
							'icon'=>t3lib_extMgm::extRelPath('pmkmp3player').'pi1/ce_wiz.gif',
							'title'=>$LANG->getLLL('pi1_title',$LL),
							'description'=>$LANG->getLLL('pi1_plus_wiz_description',$LL),
							'params'=>'&defVals[tt_content][CType]=list&defVals[tt_content][list_type]=pmkmp3player_pi1'
						);

						return $wizardItems;
					}

					/**
					 * Reads the [extDir]/locallang.xml and returns the $LOCAL_LANG array found in that file.
					 *
					 * @return	The array with language labels
					 */
					function includeLocalLang()	{
						$llFile = t3lib_extMgm::extPath('pmkmp3player').'locallang.xml';
						$LOCAL_LANG = t3lib_div::readLLXMLfile($llFile, $GLOBALS['LANG']->lang);
						
						return $LOCAL_LANG;
					}
				}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pmkmp3player/pi1/class.tx_pmkmp3player_pi1_wizicon.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pmkmp3player/pi1/class.tx_pmkmp3player_pi1_wizicon.php']);
}

?>

pi1/class.tx_pmkmp3player_pi1.php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2007-2009 Peter Klein <pmk@io.dk>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

require_once(PATH_tslib.'class.tslib_pibase.php');

/**
 * Plugin 'Flash MP3 Player' for the 'pmkmp3player' extension.
 *
 * @author	Peter Klein <pmk@io.dk>
 * @package	TYPO3
 * @subpackage	tx_pmkmp3player
 */
class tx_pmkmp3player_pi1 extends tslib_pibase {
	var $prefixId      = 'tx_pmkmp3player_pi1';		// Same as class name
	var $scriptRelPath = 'pi1/class.tx_pmkmp3player_pi1.php';	// Path to this script relative to the extension dir.
	var $extKey        = 'pmkmp3player';	// The extension key.
	var $pi_checkCHash = true;
	var $uploadDir = 'uploads/tx_pmkmp3player/';	// Upload directory
	var $genretypes = array("Blues","ClassicRock","Country","Dance","Disco","Funk","Grunge","HipHop","Jazz",
				"Metal","NewAge","Oldies","Other","Pop","R&B","Rap","Reggae","Rock","Techno",
				"Industrial","Alternative","Ska","DeathMetal","Pranks","Soundtrack","Euro-Techno",
				"Ambient","TripHop","Vocal","Jazz-Funk","Fusion","Trance","Classical","Instrumental",
				"Acid", "House", "Game", "SoundClip", "Gospel", "Noise", "Alt.Rock", "Bass", "Soul", "Punk",
				"Space", "Meditative", "InstrumentalPop", "InstrumentalRock", "Ethnic", "Gothic", "Darkwave",
				"Techno-Industrial", "Electronic", "Pop/Folk", "Eurodance", "Dream", "SouthernRock", "Comedy",
				"Cult", "GangstaRap", "Top40", "ChristianRap", "Pop/Funk", "Jungle", "NativeAmerican", "Cabaret",
				"NewWave", "Psychedelic", "Rave", "Showtunes", "Trailer", "Lo-fi", "Tribal", "AcidPunk",
				"AcidJazz", "Polka", "Retro", "Musical", "Rock'n'Roll", "HardRock", "Folk", "Folk/Rock",
				"NationalFolk","Swing", "FastFusion", "Bebob", "Latin", "Revival", "Celtic", "BlueGrass",
				"AvantGarde", "GothicRock", "ProgressiveRock", "PsychedelicRock", "SymphonicRock",
				"SlowRock", "BigBand", "Chorus", "EasyListening", "Acoustic", "Humour", "Speech", "Chanson",
				"Opera", "ChamberMusic", "Sonata", "Symphony", "BootyBass", "Primus", "PornGroove", "Satire",
				"SlowJam", "Club", "Tango", "Samba", "Folklore", "Ballad", "PowerBallad", "RhythmicSoul",
				"Freestyle", "Duet", "PunkRock", "DrumSolo", "Euro-House", "DanceHall", "Goa", "Drum&Bass",
				"Club-House", "Hardcore", "Terror", "Indie", "BritPop", "Negerpunk", "PolskPunk", "Beat",
				"ChristianGangstaRap", "HeavyMetal", "BlackMetal", "Crossover", "ContemporaryChristian",
				"ChristianRock", "Merengue", "Salsa", "ThrashMetal","Anime","JPop","SynthPop");
	var $fileList = array();			
	/**
	 * The main method of the PlugIn
	 *
	 * @param	string		$content: The PlugIn content
	 * @param	array		$conf: The PlugIn configuration
	 * @return	The content that is displayed on the website
	 */
	function main($content,$conf)	{
		$this->conf=$conf;
		$this->pi_setPiVarDefaults();
		$this->pi_loadLL();
		$this->pi_initPIflexForm();

		$TSFields = array("file","autostart","loop","bg","leftbg","rightbg","rightbghover","lefticon","righticon","righticonhover","text","slider","loader","track","border","width","height");
		$this->config = array_merge($this->getTSConfig($TSFields),$this->getFFConfig($this->cObj->data['pi_flexform']));
		
		$GLOBALS['TSFE']->additionalHeaderData[$this->extKey] = '<script type="text/javascript" src="'.t3lib_extMgm::siteRelPath($this->extKey).'res/swfobject.js"></script>
<script type="text/javascript" src="'.t3lib_extMgm::siteRelPath($this->extKey).'res/mp3player.js"></script>';
		
		// Creating valid pathes for the MP3 player
		$swfFile = t3lib_extMgm::siteRelPath($this->extKey).'res/player.swf';

		$options = '';
		while (list($k,$v) = each($this->config)) {
			switch($k) {
				case 'file' :
					$this->fileList = t3lib_div::trimExplode(',', $v, TRUE, 0);
				break;
				case 'autostart' :
				case 'loop' :
					$options.=($v!==''?','.$k.':"'.($v?'yes':'no').'"':'');
				break;
				case 'width' :
					$plwidth = ($v ? $v : '290');
				break;
				case 'height' :
					$plheight = ($v ? $v : '24');
				break;
				case 'bg':
				case 'leftbg':
				case 'rightbg':
				case 'rightbghover':
				case 'lefticon':
				case 'righticon':
				case 'righticonhover':
				case 'text':
				case 'slider':
				case 'loader':
				case 'track':
				case 'border':
					$options.=($v?','.$k.':"'.str_replace(array('aqua','black','blue','fuchsia','gray','green','lime','maroon','navy','olive','purple','red','silver','teal','yellow','white','#'),array('0x00FFFF','0x000000','0x0000FF','0xFF00FF','0x808080','0x008000','0x00FF00','0x800000','0x000080','0x808000','0x800080','0xFF0000','0xC0C0C0','0x008080','0xFFFF00','0xFFFFFF','0x'),$v).'"':'');
				break;
				default :
				//$options.=($v?','.$k.':"'.$v.'"':'');
			
			}
		}
		$content = $this->cObj->cObjGetSingle($conf['beforeAllObj'], $conf['beforeAllObj.']);
		
		// Loop over MP3s
		// In this loop we have to fill $this->cObj->data
		$playerSubID = 1;
		foreach ($this->fileList as $file) {
			$this->cObj->data =array_merge($this->cObj->data,$this->getID3v11($this->uploadDir . $file));
			$this->cObj->data['file'] = $file;
			$this->cObj->data['filesize'] = filesize($this->uploadDir . $file);
			$playerid = ''.($this->cObj->data['uid'] * 1000) + $playerSubID;
			$fileOptions = 'playerID:"'.$playerid.'"'.$options.($file?',soundFile:"'.str_replace(PATH_site,t3lib_div::getIndpEnv('TYPO3_SITE_URL'),t3lib_div::getFileAbsFileName($this->uploadDir . $file)).'"':'');
			$content .= $this->cObj->cObjGetSingle($conf['beforeObj'], $conf['beforeObj.']);
			$content.= '<p id="'.$this->extKey.$playerid.'"><a href="http://www.macromedia.com/go/getflashplayer">'.$this->pi_getLL('getFlash1').'</a> '.$this->pi_getLL('getFlash2').'</p>
<script type="text/javascript">
  swfobject.embedSWF("'.$swfFile.'","'.$this->extKey.$playerid.'", "'.$plwidth.'", "'.$plheight.'", "7.0.0", "expressInstall.swf",{'.$fileOptions.'},{wmode:"transparent",bgcolor:"#FFFFFF"},{id:"audioplayer'.$playerid.'"});
</script>';
			$content .= $this->cObj->cObjGetSingle($conf['afterObj'], $conf['afterObj.']);
			$playerSubID++;
		}
		$content .= $this->cObj->cObjGetSingle($conf['afterAllObj'], $conf['afterAllObj.']);
		return $this->pi_wrapInBaseClass($content);
	}
	
	/**
	 * Get config data from Flexform
	 *
	 * @param	string		$ffObj: Flexform object
	 * @return	array		config array
	 */
	function getFFConfig($ffObj) {
		$config = array();
		foreach($ffObj['data'] as $sheet => $arr) {
			foreach($arr['lDEF'] as $field => $v) {
				$val = $this->pi_getFFvalue($ffObj,$field,$sheet);
				if ($val!=='') $config[$field] = $val;
			}
		}
		return $config;
	}

	function getTSConfig($confFields) {
		while (list(,$field) = each($confFields)) {
			$val = $this->cObj->stdWrap($this->conf[$field], $this->conf[$field.'.']);
			if ($val!=='') $config[$field] = $val;
		}
		return $config;
	}

	function getID3v11($file) {
		if (!file_exists($file)) return array();
		if (!$file=fopen($file, "rb")) return array();
		fseek($file, -128, SEEK_END);
		$tagheader = fgets($file, 129);
		if (substr($tagheader,0,3)=='TAG') {
			$ID3v11['title'] = trim(substr($tagheader,3,30));
			$ID3v11['artist'] = trim(substr($tagheader,33,30));
			$ID3v11['album'] = trim(substr($tagheader,63,30));
			$ID3v11['year'] = substr($tagheader,93,4);
			$ID3v11['comment'] = trim(substr($tagheader,97,28));
			$ID3v11['track'] = hexdec(bin2hex(substr($tagheader,126,1)));
			$ID3v11['genrenum'] =hexdec(bin2hex(substr($tagheader,127,1)));
			$ID3v11['genre'] = @$this->genretypes[$ID3v11['genrenum']];
			return $ID3v11;
		}
		else return array();
	}
}



if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pmkmp3player/pi1/class.tx_pmkmp3player_pi1.php'])	{
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/pmkmp3player/pi1/class.tx_pmkmp3player_pi1.php']);
}

?>
